flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java

```diff
@@ -1142,14 +1146,27 @@ protected Configuration configuration() {
 
     public static List<TestSpec> constructedTypes() {
         return Arrays.asList(
-                // https://issues.apache.org/jira/browse/FLINK-17321
-                // MULTISET
-                // MAP
+                CastTestSpecBuilder.testCastTo(MAP(STRING(), STRING()))
+                        .fromCase(MAP(FLOAT(), DOUBLE()), null, null)
+                        .fromCase(
+                                MAP(INT(), INT()),
+                                Collections.singletonMap(1, 2),
+                                Collections.singletonMap("1", "2"))
```

Comment:
:thinking: I'm surprised that this works without changing anything in `LogicalTypeCasts`... Can you add a test case for failure? For example:

```
CastTestSpecBuilder
  .testCastTo(MAP(STRING(), STRING()))
  .fail(MAP(STRING(), ROW(INT())), whateverValue)
```

Same for ROW to ROW and ARRAY to ARRAY. Just take whatever inner invalid tuple and see what exception you get. If it fails within the single `CastRule` implementations, then something is wrong with `LogicalTypeCasts`.

Reply:
I didn't get, should it fail?
```sql
SELECT CAST(MAP['a', row(1)] AS MAP<STRING, STRING>);
```
I thought it should return something like that
```
{a=(1)}

Reply:
It should not fail, the goal of this issue is to allow such casting. The thing is that I'm actually surprised it doesn't fail without touching `LogicalTypeCasts`, which is the class that checks valid/invalid casts.

EDIT: I've done some debugging, `LogicalTypeCasts#supportsConstructedCasting` does the job and just checks for the equality of the type root and for the castability of the children. So it works :smile: